File: bestia-client/src/Bnet/Message/Entity/LogoutIntentComponentSMSG.cs
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Entity
{
  /// <summary>
  /// Server-driven logout countdown for the local player's master. Re-sent periodically; the client
  /// interpolates locally between updates. Cancellation arrives as this same message with
  /// Removed = true. It IS an EntitySMSG (carries EntityId), but is routed straight to the logout UI
  /// rather than through the generic per-entity dispatch — see the ordering note in
  /// connection_manager.gd's message dispatch.
  /// </summary>
  [GlobalClass]
  public partial class LogoutIntentComponentSMSG : EntitySMSG
  {
    [Export] public float RemainingSeconds { get; set; }
    [Export] public bool Removed { get; set; }

    public static LogoutIntentComponentSMSG FromProto(global::Bnet.LogoutIntentSMSG proto)
    {
      return new LogoutIntentComponentSMSG()
      {
        EntityId = proto.EntityId,
        RemainingSeconds = proto.RemainingSeconds,
        Removed = proto.Removed
      };
    }

    public override string ToString()
    {
      return $"LogoutIntentSMSG(EntityId={EntityId}, RemainingSeconds={RemainingSeconds}, Removed={Removed})";
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/ecs/logout/LogoutIntentComponentSMSG.kt
package net.bestia.zone.ecs.logout

import net.bestia.bnet.proto.EnvelopeProto
import net.bestia.bnet.proto.LogoutIntentSmsgProto
import net.bestia.zone.message.EntitySMSG

/**
 * Owner-only sync of a pending logout countdown on the master entity. Produced by
 * [LogoutIntent.toEntityMessage]; re-sent periodically so the client countdown stays corrected.
 */
data class LogoutIntentComponentSMSG(
  override val entityId: Long,
  val remainingSeconds: Float,
  val removed: Boolean = false
) : EntitySMSG {

  override fun toBnetEnvelope(): EnvelopeProto.Envelope {
    val proto = LogoutIntentSmsgProto.LogoutIntentSMSG.newBuilder()
      .setEntityId(entityId)
      .setRemainingSeconds(remainingSeconds)
      .setRemoved(removed)
      .build()

    return EnvelopeProto.Envelope.newBuilder()
      .setCompLogoutIntent(proto)
      .build()
  }
}
